frontend/src/components/widgets/DataFrame/EditingState.test.ts

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { GridCell, TextCell, GridCellKind } from "@glideapps/glide-data-grid"
+
+import {
+  BaseColumnProps,
+  TextColumn,
+} from "src/components/widgets/DataFrame/columns"
+
+import EditingState from "./EditingState"
+
+const MOCK_TEXT_CELL_1: TextCell = {
+  kind: GridCellKind.Text,
+  displayData: "foo",
+  data: "foo",
+  allowOverlay: true,
+}
+
+const MOCK_TEXT_CELL_2: TextCell = {
+  kind: GridCellKind.Text,
+  displayData: "foo",
+  data: "foo",
+  allowOverlay: true,
+}
+
+describe("EditingState class", () => {
+  it("allows to set edited cells", () => {
+    const NUM_OF_ROWS = 3
+    const editingState = new EditingState(NUM_OF_ROWS)
+    editingState.setCell(0, 0, MOCK_TEXT_CELL_1)
+    expect(editingState.getCell(0, 0)).toEqual(MOCK_TEXT_CELL_1)
+    expect(editingState.getCell(1, 1)).toEqual(undefined)
+
+    // Overwrite cell
+    editingState.setCell(0, 0, MOCK_TEXT_CELL_2)
+    expect(editingState.getCell(0, 0)).toEqual(MOCK_TEXT_CELL_2)
+  })
+
+  it("allows to add rows", () => {
+    const NUM_OF_ROWS = 3
+    const editingState = new EditingState(NUM_OF_ROWS)
+
+    const rowCells: Map<number, GridCell> = new Map()
+    rowCells.set(0, MOCK_TEXT_CELL_1)
+    rowCells.set(1, MOCK_TEXT_CELL_2)
+
+    // Add a row and check values
+    editingState.addRow(rowCells)
+    expect(editingState.getNumRows()).toEqual(NUM_OF_ROWS + 1)
+    expect(editingState.isAddedRow(2)).toEqual(false)
+    expect(editingState.isAddedRow(3)).toEqual(true)
+    expect(editingState.getCell(0, 3)).toEqual(MOCK_TEXT_CELL_1)
+    expect(editingState.getCell(1, 3)).toEqual(MOCK_TEXT_CELL_2)
+
+    // Add another row
+    editingState.addRow(rowCells)
+    expect(editingState.getNumRows()).toEqual(NUM_OF_ROWS + 2)
+    expect(editingState.getCell(0, 4)).toEqual(MOCK_TEXT_CELL_1)
+    expect(editingState.getCell(1, 4)).toEqual(MOCK_TEXT_CELL_2)
+  })
+
+  it("allows to delete rows", () => {
+    const NUM_OF_ROWS = 3
+    const editingState = new EditingState(NUM_OF_ROWS)
+
+    // Delete first row
+    editingState.deleteRow(0)
+    expect(editingState.getNumRows()).toEqual(NUM_OF_ROWS - 1)
+
+    // The current row 0 should be the original row 1
+    expect(editingState.getOriginalRowIndex(0)).toEqual(1)
+  })
+
+  it("allows to delete rows", () => {
```

Comment:
Nit: "allows to delete rows" -> "allows to delete rows after adding rows"
I think when anyone runs this test and it fails, it will be more obvious especially if the test above also fails.